packages/core/src/transport.ts

```diff
@@ -46,33 +46,56 @@ export class Batch<T> {
   }
 
   add(message: T) {
+    this.addOrUpdate(message)
+  }
+
+  upsert(message: T, key: string) {
+    this.addOrUpdate(message, key)
+  }
+
+  remove(index: number) {
```

Comment:
When you are removing an event, you don't update the `bufferIndexByKey` indexes, so they become out of sync. See this test (failing with the current implementation):

```js
    batch.upsert({ message: '1' }, 'a')
    batch.upsert({ message: '1' }, 'b')
    batch.upsert({ message: '2' }, 'a')
    batch.upsert({ message: '3' }, 'b')
    batch.flush()

    expect(transport.send).toHaveBeenCalledWith(
      '{"foo":"bar","message":"2"}\n{"foo":"bar","message":"3"}',
      jasmine.any(Number)
    )
```

I think having an `events` structure like this:
```ts
private events = Array<{event: string, key: string | undefined}>
```
and changing the `remove` method to accept a key instead of an index would be better/safer.

Reply:
Nice catch! I'll look into that

Reply:
I've tried a slighly different strategy to avoid to iterate over the buffer at each upsert, let me know what you think.

Reply:
Okay, should be fine :)